design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+Errors may also end up traveling through several different components written in different languages before becoming a failure.  To assist with diagnosing these failures, xlang will allow a projection to associate additional diagnostic information to the existing information as an error travels through it.  This will help a developer get a complete picture of the failure which wouldn't be available with just the information collected from the language that initially returned the error.  An example of this is the stack trace.  A language will typically only be aware of the stack within its language boundary and thereby the diagnostic information collected at the initial error would only have a partial stack.  The following additional information can be associated with the error as it travels through projections:
+
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+The diagnostic information will be stored in a COM object (error info) implementing the IXlangErrorInfo interface.  As with any COM object, it will be reference counted and will implement IUnknown.  This will allow it to be easily accessed by projections written in different languages while still being able to ensure the cleanup of it.
+
+## Handling language errors
+
+Producing projections may encounter a language error when projecting an API call to a component.  The error will either be from the translation of the call to the component's language or be from the component itself as a result of the call.  Either way, upon encountering an error, the projection should determine whether the error is new or is a propagation of a previous xlang error.  Strategies on doing this will be discussed in a later section.
+
+If the error is new, the projection should translate the error to an equivalent xlang error or to xlang_fail if there is no equivalent mapping.  After translating the error, the projection should call xlang's error origination API, xlang_originate_error, to store the diagnostic information collected by the language for the error.  The API will create an error info with the provided diagnostic information and store it on a thread level storage.  The created error info will remain on the thread level storage until it is retrieved by the consuming projection.
+
+xlang_originate_error will be defined in the PAL and will take the following parameters:
+- a xlang_result with the translated xlang error for the language specific error that occurred
+- a xlang_string with the language specific error that occurred
+- a xlang_string with the error message
+- a xlang_string with the error stack if available
+- a xlang_string with an identifier for the projection
+- an IUnknown pointer with any additional language specific information (can be null)
```

Comment:
As per earlier comment, this should be IXlangObject

Reply:
Thought about this a bit.  I think we want to provide the capability for a projection to pass a non-projected language object to this field without needing to worry about wrapping it in an IXlangObject or implement that interface on it.  An example of this would be when a projection wants to store information to re-construct their exception later.  That object shouldn't need to be a projected type as it is intended for only them to retrieve it later.  In addition, I don't expect a projected component / consumer to be looking at this parameter, but rather projections exposing it in the language exception if it implements IXlangExtendedErrorInfo or something they recognize.  Due to that, leaving it as IUknown which is the xlagn equivalent for non projected types.

Reply:
Leaving it as IUnknown is a mistake in my opinion. In practically every case in WinRT, when we decided to "optimize" by using IUnknown instead of IInspectable (e.g. delegates), it turns out to be a mistake.

Encoding policy decisions ("dont expect") into the type system is limiting and often wrong over time.

IMO this should be IXlangObject so it's something that can be represented / manipulated within the xlang type system.

Reply:
I understand what you are saying with the delegates in WinRT.  But, the current model proposed is an app calling a xlang component shouldn't directly be interacting with the error info object.  The projection for the app's language should take the info in the error info object and expose it as part of the language error for that projection.  This means that the members we store as part of the error info object don't need to be projected, but we need to allow a projection to take that info and put its details as part of its language error representation which we do and talk about in the design note.  If we do end up seeing a need in the future where the error info object needs to be projected to the app, we can end up adding another interface to allow for that.

Also a projection is free to stored a projected object in this parameter, but it would need to expose it as part of its own interface.